Add forward AD support for miscellanous operators

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67045
* #67041
* __->__ #67040
* #66294
* #66293
* #66292

 - Add forward AD formulas for:
   - Public API `relu`, `stack`, `neg`, `binary_cross_entropy_with_logits`
   - Non-public-API: `threshold_backward`
 - We rely on OpInfos to test all them except `binary_cross_entropy_with_logits`.
 - Temporary: applies a fix for non-differentiable ops, but will be rebased away later

soulitzer committed:
Add forward AD support for miscellanous operators

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for miscellanous operators"


 - Add forward AD formulas for:
   - Public API `relu`, `stack`, `neg`, `binary_cross_entropy_with_logits`
   - Non-public-API: `threshold_backward`
 - We rely on OpInfos to test all them except `binary_cross_entropy_with_logits`.
 - Temporary: applies a fix for non-differentiable ops, but will be rebased away later
 

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for miscellanous operators"


 - Add forward AD formulas for:
   - Public API `relu`, `stack`, `neg`, `binary_cross_entropy_with_logits`
   - Non-public-API: `threshold_backward`
 - We rely on OpInfos to test all them except `binary_cross_entropy_with_logits`.
 - Temporary: applies a fix for non-differentiable ops, but will be rebased away later
 

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD support for miscellanous operators"


 - Add forward AD formulas for:
   - Public API `relu`, `stack`, `neg`, `binary_cross_entropy_with_logits`
   - Non-public-API: `threshold_backward`
 - We rely on OpInfos to test all them except `binary_cross_entropy_with_logits`.
 - Temporary: applies a fix for non-differentiable ops, but will be rebased away later
 

[ghstack-poisoned]

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1642,6 +1644,7 @@
 
 - name: relu(Tensor self) -> Tensor
   self: threshold_backward(grad, result, 0)
+  result: threshold_backward(self_t, self_p, 0)
```
auto_element_wise

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1055,6 +1055,7 @@
 
 - name: neg(Tensor self) -> Tensor
   self: grad.neg()
+  result: self_t.neg()
```
auto_element_wise

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -2053,6 +2056,7 @@
 - name: threshold_backward(Tensor grad_output, Tensor self, Scalar threshold) -> Tensor
   grad_output: threshold_backward(grad, self, threshold)
   self: zeros_like(grad)
+  result: zeros_like(self_t) + threshold_backward(grad_output_t, self_p, threshold)
```
What is the zeros_like for here?

soulitzer replied:
the one in the native namespace?

albanD replied:
Ho I meant why are you adding 0s to something? Why not just have the something?

soulitzer replied:
Ah it should be for when the self has tangent but grad_output does not, then LHS is batched but RHS is not, so we'd like to preserve the batchedness.

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1332,6 +1342,37 @@ Tensor binary_cross_entropy_with_logits_target_backward(const Tensor& grad_outpu
+  if (pos_weight.defined()) {
+    // pos_weight need to be broadcasted, thus mul(target) is not inplace.
+    auto t = pos_weight.mul(target_p);
+    grad_input = input_t.mul_(t.add(1).sub_(target_p).mul_(input_p.sigmoid()).sub_(t));
```
The tangent should not be modified in the forward was not inplace.

soulitzer committed:
Update on "Add forward AD support for miscellanous operators"


 - Add forward AD formulas for:
   - Public API `relu`, `stack`, `neg`, `binary_cross_entropy_with_logits`
   - Non-public-API: `threshold_backward`
 - We rely on OpInfos to test all them except `binary_cross_entropy_with_logits`.
 - Temporary: applies a fix for non-differentiable ops, but will be rebased away later
 

[ghstack-poisoned]

albanD reviewed (approved):
LGTM as long as the CI is green

soulitzer commented:
Closing in favor of https://github.com/pytorch/pytorch/pull/67820
